source/simulator/solver_schemes.cc

```diff
@@ -564,86 +576,113 @@ namespace aspect
+             * is met, or the line search iteration limit is reached.
+             */
+            LinearAlgebra::BlockVector backup_linearization_point = current_linearization_point;
 
-            assemble_newton_stokes_matrix = rebuild_stokes_preconditioner = false;
-            rebuild_stokes_matrix = true;
 
-            assemble_stokes_system();
+            double test_velocity_residual = 0;
+            double test_pressure_residual = 0;
+            double lambda = 1;
+            double alpha = 1e-4;
+            unsigned int line_search_iteration = 0;
 
-            test_velocity_residual = system_rhs.block(introspection.block_indices.velocities).l2_norm();
-            test_pressure_residual = system_rhs.block(introspection.block_indices.pressure).l2_norm();
-            test_residual = std::sqrt(test_velocity_residual * test_velocity_residual
-                                      + test_pressure_residual * test_pressure_residual);
 
-            if (test_residual < (1.0 - alpha * lambda) * residual
-                ||
-                line_search_iteration >= parameters.max_newton_line_search_iterations
-                ||
-                use_picard)
-              {
-                pcout << "      Relative nonlinear residual (total Newton system) after nonlinear iteration " << nonlinear_iteration+1
-                      << ": " << residual/initial_residual << ", norm of the rhs: " << test_residual
-                      << ", newton_derivative_scaling_factor: " << newton_handler->get_newton_derivative_scaling_factor() << std::endl;
-                break;
-              }
-            else
+            /**
+             * Do the loop for the line search. Even when we
+             * don't do a line search we go into this loop
+             */
+            do
               {
-
-                pcout << "   Line search iteration " << line_search_iteration << ", with norm of the rhs "
-                      << test_residual << " and going to " << (1.0 - alpha * lambda) * residual
-                      << ", relative residual: " << residual/initial_residual << std::endl;
-
-                /**
-                 * The line search step was not sufficient to decrease the residual
-                 * enough, so we take a smaller step to see if it improves the residual.
-                 */
-                lambda *= (2.0/3.0);// TODO: make a parameter out of this.
+                // Reset the current linearization point and the search direction
+                current_linearization_point = backup_linearization_point;
+                LinearAlgebra::BlockVector search_direction = solution;
+
+                // Scale the search direction down and add it to the current linearization point.
+                search_direction *= lambda;
+
+                current_linearization_point.block(introspection.block_indices.pressure) += search_direction.block(introspection.block_indices.pressure);
```

Comment:
you can use ``current_linearization_point.block(..).add(lambda, solution.block(..))`` and remove ``search_direction``

Reply:
this is indeed a lot cleaner, thanks!

Reply:
on second thought, it doesn't work in debug mode because of the ghost vectors of the current_linearization_point. I get an assert: `You are trying an operation on a vector that is only allowed if the vector has no ghost elements, but the vector you are operating on does have ghost elements. Specifically, vectors with ghost elements are read-only and cannot appear in operations that write into these vectors.`

Reply:
@tjhei is there a way around this or should I just revert to what it was?